### ISSUE TITLE
[question] Enforcing a minimum C++ standard for consumers of a package while remaining compiler and platform independent

### ISSUE THREAD

**Issue by tjwrona** [conan-io/conan#7772]
I have a library that uses some C++17 language features and I am in the process of trying to package it up using Conan so it can be easily consumed by any build system. Because some of the header files in my library use C++17 features, the consumer of the library is also required compile their code with at least a minimum standard of C++17 otherwise compilation will fail.

In the `package_info` section of my conanfile I can tell it to use the proper compiler flag:

```python
self.cpp_info.cxxflags = ["-std=c++17"]
```

But this has two issues:

1. This compiler flag is not recognized by all compilers. For example this works with gcc, but if I try to use MSVC I will get an error `warning D9002: ignoring unknown option '-std=c++17'` and the code will fail to compile.

2. This locks the consumer of the package into using C++17 when a newer standard would work. What happens when the consumer decides they want to use my library but they want to compile their code using C++20?
________

With a pure CMake project that doesn't use Conan this is handled quite elegantly by just adding:

```cmake
target_compile_features(<target> PUBLIC cxx_std_17)
```

This will automatically pass the minimum standard requirement to the consumer of the target AND should they decide to compile their code with a newer standard (like C++20) CMake will account for that and the entire library will be compiled with the newer standard instead.
_______

While the pure CMake solution is nice, I can't rely on it and still have my library be consumable by users who decide to use a different build system other than CMake. How can I solve this problem with Conan?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by Johnnyxy** [comment]
Hi there,
have a look here, maybe this could help you to enforce a minimum C++ standard.
[https://docs.conan.io/en/latest/reference/tools.html#tools-check-min-cppstd](https://docs.conan.io/en/latest/reference/tools.html#tools-check-min-cppstd)

**Comment by kralicky** [comment]
You can also use the `settings` field to restrict your package to certain c++ standards, something like this:

```python
settings = {
    "os": ["Linux", "Windows"],
    "compiler": {"gcc": {"cppstd": [17, 20]},
                 "Visual Studio": {"cppstd": [17, 20]}},
    "arch": None,
    "build_type": None
}
```
You can see all the possible settings in ~/.conan/settings.yml.

[Relevant docs here](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#settings)

**Comment by tjwrona** [comment]
Thanks both of you! I like the `tools.check_min_cppstd` option a bit more because it seems more open ended so it will continue to work with the next standard after C++20.

The only issue I have with it is that you have to pass the `compiler.cppstd` as an option to `conan install` and it won't automatically pick up that you need at least that standard and do it for you. The pure CMake approach is more automatic because all you have to do is consume the target and it works out of the box.

**Comment by jgsogo** [comment]
Hi! To be honest, we don't have a canonical solution, but it is something we've been thinking about and we need to come up with a plan for Conan 2.0. Probably a good reading is around issue https://github.com/conan-io/conan/issues/6157 and other issues linked to it.

To summarize a little bit: the way to handle C++ standard in Conan should be using `compiler.cppstd` setting, but we don't have a mechanism to enforce a setting value for the consumer of a library. If your recipe uses `tools.check_min_cppstd` (or a custom implementation) it can raise an error if the provided profile doesn't define a value (high enough) for `compiler.cppstd`, but Conan won't assign that value based on some attribute declared in the recipes that are part of the full-graph.

Nevertheless, C++ standard is a bit tricky, stable implementations are binary compatible, you only need it in the consumer if the headers of your libraries expose some new C++ feature. With this in mind, it depends on the API of your library if you should use `tools.check_min_cppstd` in the `configure()` method (it is evaluated always) or only in the `build()` method (that C++ standard is only required if you are going to build the library). And then, depending on the ABI compatibility of your library, you might want to remove `compiler.cppstd` from the package-id computation: if all the binaries are compatible, maybe you just want to generate one of them and use it regardless of the value of `compiler.cppstd`.

Thinking about future features in Conan: a way to declare API and ABI compatibility based on the C++ standard might be added in the future, but probably not before Conan 2.0.

And of course, we need to have in mind, that using `compiler.cppstd` will pass the corresponding flags to the build system, but if it is the `CMakeLists.txt` the one that defines the C++ standard to use (typical `set_target_properties(<target> PROPERTIES CXX_STANDARD 11`)), Conan won't know anything about it.

I'm so sorry, I don't have a closed answer for this, we are working hard to improve C++ ecosystem but there are lots of things to do 😄 

Thanks!

**Comment by tjwrona** [comment]
Hi @jgsogo, thanks for the in-depth response. Regarding this part:

>Nevertheless, C++ standard is a bit tricky, stable implementations are binary compatible, you only need it in the consumer if the headers of your libraries expose some new C++ feature. With this in mind, it depends on the API of your library if you should use tools.check_min_cppstd in the configure() method (it is evaluated always)

I do expose some new C++ features in the headers for my library so the consumer does need to be compiled using at least C++17. I put `tools.check_min_cppstd` in the `configure()` method of my conanfile.py for the library I am distributing, but it doesn't seem to be evaluated when I consume the library, only when I create it.

Basically my project that consumes the library ends up failing to compile but it doesn't fail at the `configure()` stage, it fails during actual compilation.

**Comment by tjwrona** [comment]
@jgsogo, I also have this header only library that requires the consumer to use a C++17 compatible compiler:

https://github.com/tnt-coders/cpp-benchmark

There doesn't seem to be any way in the `conanfile` to enforce that the consumer uses a C++17 compiler until they just try and build their code and it fails. Putting the `tools.check_min_cppstd` check in the `configure()` method does not achieve this.

**Comment by kralicky** [comment]
@tjwrona1992 I think that for your use case, you should use `settings` to restrict compiler versions to compilers which support C++17, instead of restricting the C++ version itself.

**Comment by tjwrona** [comment]
@cobalt77 is there a way to restrict the setting to C++17 or newer without explicitly listing each version? I don't want the recipe to break when the next version of C++ comes out and someone tries to use a compiler with C++23 or something.

**Comment by Johnnyxy** [comment]
You could restrict it in the ```configure``` function.

```python
def configure(self):
    if self.settings.compiler.cppstd in [None, "11", "14"]:
        raise ConanInvalidConfiguration("C++ standard less than 17 not supported")
```

**Comment by tjwrona** [comment]
@Johnnyxy I already tried using `tools.check_min_cppstd` in the `configure` function which should have the same effect. The issue is that `configure` function doesn't get called if I'm simply consuming the package and it has already been installed.

I can install the package using `conan install .. --settings compiler.cppstd=17`

But then when I actually compile the project that uses the package with CMake there is no way to have Conan enforce that the consuming project be compiled with C++17. The code that is consuming the package is failing to compile because it is using header files from the package that require C++17. The package itself builds and installs fine.

**Comment by Johnnyxy** [comment]
The consuming project has to specify ```settings.compiler.cppstd``` on ```conan install ... -s compiler.cppstd=20```.
At that point your consumed package gets invoked and it can verify that one cannot install it with a C++ standard less than 17.
This means your consumed package enforces that it cannot be installed with a Conan C++ standard setting less than 17. But as you stated this has no impact on the build system itself.

I think you have to differentiate here. Conan is not for building. Conan is a package/dependency management tool. I does not do the work of a toolchain (compiling, linking, packaging, ...). If the toolchain or code needs something that restricts its usage the developer of the code/build script (e.g. CMake, makefile) has to provide some means for that. Then one can instruct the Conan package to provide informations that the toolchain can then decide upon if it should/can compile.
As @jgsogo stated (https://github.com/conan-io/conan/issues/7772#issuecomment-702053941), Conan does not know what the build system does. E.g. you could run CMake to generate makefiles on-the-fly and then run them or invoke some extra build tool or whatever crazy things some developers can think of.

Conan expects certain outcomes after its functions. For the ```build``` function Conan expects that all binaries are available on success. The ```package``` function provides all binaries in the package folder (on success). Nothing more or less.

I made a mistake at the code example in my last post. I corrected it.

**Comment by tjwrona** [comment]
@Johnnyxy okay I guess that makes sense... It's a bit unfortunate because if I import the CMake targets for my package by using the `cmake_paths` generator and then link to that imported target CMake will automatically pick up the requirement for C++17 and everything compiles okay.

This means that I get different behavior using the `cmake` generator than I do if I use the `cmake_paths` generator.

In theory this could be fixed if you could somehow embed the same information for the C++ standard into the target that Conan generates with the `cmake` generator.

**Comment by Johnnyxy** [comment]
If you are using CMake then there is some integration which can help you to achieve your goal.
In theory for every CMake target that Conan generates the required information can be set. 

```cmake_find_package*``` as well as ```cmake*``` (see [here](https://docs.conan.io/en/latest/reference/generators/cmake.html#conan-basic-setup) and [here](https://docs.conan.io/en/latest/reference/generators/cmake.html#conan-set-std)) could help in this regard if they are not doing it already (@jgsogo).



But I doubt the ```cmake_paths``` itself could provide you with any additional information at the moment (see [here](https://docs.conan.io/en/latest/reference/generators/cmake_paths.html)). It is designed to provide only CMake search paths.

**Comment by tjwrona** [comment]
That's the thing, the `cmake_paths` generator is the only one that actually works out of the box because although it only sets the search paths, it allows Conan to find and use my ACTUAL installed CMake target that has the C++ standard embedded into it. This makes anything that links to the target also use the correct C++ standard.

For `cmake_find_package` and `cmake` that information does not get embedded into the target so CMake doesn't know what standard to use when you link to the Conan target.

**Comment by Johnnyxy** [comment]
There you have it. You are using your own build system files (CMakeLists.txt, FindPACKAGE.cmake, ...) without relying on anything from Conan.
At the moment you already have those functionality and have to reproduce it with the ```configure```-check to ensure that the Conan recipe does these checks too. I agree to the argument that this would be kind of redundant.

Maybe the new [Toolchain-feature](https://docs.conan.io/en/latest/creating_packages/toolchains/cmake.html) could help you but it is still experimental.

**Comment by tjwrona** [comment]
I'm okay with it if I have to redundantly define the C++ standard in the conanfile in order to have the Conan target embed that information into it, but it doesn't look like there is any way to embed that information into a Conan target at all so any targets generated by Conan cannot automatically pass the C++ standard to the consuming project.

For now I will probably just start using `cmake_paths` as my preferred generator since that pulls in the actual CMake target and passes along the C++ standard information to the consumer. I'll make note in the README file that other generators (`cmake` and `cmake_find_package`) will not capture the C++17 requirement when consuming the package so using `cmake_paths` is preferred otherwise you will need to manually specify that you are using C++17 in the consuming project.

I will take a look at the toolchain stuff though and see if that provides any more useful options.

**Comment by AbrilRBS** [comment]
Closing as solved in Conan 2, where `compiler.cppstd` is better modeled and supports these operations in a native way
